pkg/dss/cockroach/store.go

```
@@ -18,6 +76,188 @@ func (s *Store) Close() error {
+	const (
+		subscriptionStatusQuery = `
+		INSERT INTO
+			subscriptions_status
+		VALUES
+			($1, $2, $3)
+		RETURNING
+			*
+		`
+		subscriptionQuery = `
+		INSERT INTO
+			subscriptions
+		VALUES
+			(gen_random_uuid(), $1, $2, $3, $4, $5, transaction_timestamp())
+		RETURNING
+			*`
+	)
+
+	sr := subscriptionsRow{
+		owner: subscription.GetOwner(),
+		url:   subscription.Callbacks.GetIdentificationServiceAreaUrl(),
+	}
+
+	if ts := subscription.GetBegins(); ts != nil {
+		begins, err := ptypes.Timestamp(ts)
+		if err != nil {
+			return nil, err
+		}
+		sr.beginsAt.Time = begins
+		sr.beginsAt.Valid = true
+	}
+
+	if ts := subscription.GetExpires(); ts != nil {
+		expires, err := ptypes.Timestamp(ts)
+		if err != nil {
+			return nil, err
+		}
+		sr.expiresAt.Time = expires
+		sr.expiresAt.Valid = true
+	}
+
+	tx, err := s.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := sr.scan(tx.QueryRowContext(
+		ctx,
+		subscriptionQuery,
+		sr.owner,
+		sr.url,
+		sr.typesFilter,
+		sr.beginsAt,
+		sr.expiresAt,
+	)); err != nil {
+		return nil, multierr.Combine(err, tx.Rollback())
+	}
+
+	ssr := subscriptionsStatusRow{
+		subscriptionID:    sr.id,
+		notificationIndex: int64(subscription.NotificationIndex),
+	}
+
+	if err := ssr.scan(tx.QueryRowContext(
+		ctx,
+		subscriptionStatusQuery,
+		ssr.subscriptionID,
+		ssr.notificationIndex,
+		ssr.lastUsedAt,
+	)); err != nil {
+		return nil, multierr.Combine(err, tx.Rollback())
+	}
+
+	result := &dspb.Subscription{
+		Id:    sr.id,
+		Owner: sr.owner,
+		Callbacks: &dspb.SubscriptionCallbacks{
+			IdentificationServiceAreaUrl: sr.url,
+		},
+		NotificationIndex: int32(ssr.notificationIndex),
+	}
+
+	if sr.beginsAt.Valid {
+		ts, err := ptypes.TimestampProto(sr.beginsAt.Time)
+		if err != nil {
+			return nil, multierr.Combine(err, tx.Rollback())
+		}
+		result.Begins = ts
+	}
+
+	if sr.expiresAt.Valid {
```

if we intentionally didn't insert a timestamp then i believe this will be invalid. 

I think we should throw an error if we inserted a timestamp and it's not valid. This might lend itself to throwing an error in scan and getting rid of the Valid concept

---

Ah, I think I should add comments to `nullTime` I guess. Its semantics follow https://golang.org/pkg/database/sql/#NullString, i.e., `Valid` indicates that there was a non-`NULL` value in the database. It does not indicate whether the timestamp in the database is valid. We couldn't have inserted an invalid timestamp either, as the DB would have bailed out. We *should* probably establish table-CHECKs, though, that ensure `expires_at > begins_at` if both are not `NULL`.

---

We could model the relevant columns `NOT NULL`, but the API documentation indicates that both are optional.